pkg/flavor/flavor.go

```diff
@@ -30,27 +34,115 @@ func (c Conflicts) String() string {
+	preferenceStore        cache.Store
+	clusterPreferenceStore cache.Store
 }
 
 var _ Methods = &methods{}
 
-func NewMethods(flavorStore, clusterFlavorStore cache.Store) Methods {
+func NewMethods(flavorStore, clusterFlavorStore, preferenceStore, clusterPreferenceStore cache.Store) Methods {
 	return &methods{
-		flavorStore:        flavorStore,
-		clusterFlavorStore: clusterFlavorStore,
+		flavorStore:            flavorStore,
+		clusterFlavorStore:     clusterFlavorStore,
+		preferenceStore:        preferenceStore,
+		clusterPreferenceStore: clusterPreferenceStore,
 	}
 }
 
-func (m *methods) ApplyToVmi(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
-
+func (m *methods) ApplyToVmi(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, preferenceSpec *flavorv1alpha1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
 	var conflicts Conflicts
 
-	conflicts = append(conflicts, applyCpu(field, flavorSpec, vmiSpec)...)
+	if flavorSpec != nil {
+		conflicts = append(conflicts, applyCpu(field, flavorSpec, preferenceSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyMemory(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyIOThreadPolicy(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyLaunchSecurity(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyGPUs(field, flavorSpec, vmiSpec)...)
+		conflicts = append(conflicts, applyHostDevices(field, flavorSpec, vmiSpec)...)
+	}
+
+	if preferenceSpec != nil {
+		// By design Preferences can't conflict with the VMI so we don't return any
+		applyDevicePreferences(preferenceSpec, vmiSpec)
+		applyFeaturePreferences(preferenceSpec, vmiSpec)
+		applyFirmwarePreferences(preferenceSpec, vmiSpec)
+		applyMachinePreferences(preferenceSpec, vmiSpec)
+		applyClockPreferences(preferenceSpec, vmiSpec)
+	}
 
 	return conflicts
+}
+
+func (m *methods) FindPreferenceSpec(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachinePreferenceSpec, error) {
+
+	var err error
+	var preference *flavorv1alpha1.VirtualMachinePreference
+	var clusterPreference *flavorv1alpha1.VirtualMachineClusterPreference
+
+	if vm.Spec.Preference == nil {
+		return nil, nil
+	}
+
+	switch strings.ToLower(vm.Spec.Preference.Kind) {
+	case apiflavor.SingularPreferenceResourceName, apiflavor.PluralPreferenceResourceName:
+		preference, err = m.findPreference(vm)
+	case apiflavor.ClusterSingularPreferenceResourceName, apiflavor.ClusterPluralPreferenceResourceName:
+		clusterPreference, err = m.findClusterPreference(vm)
+	default:
+		err = fmt.Errorf("got unexpected kind in PreferenceMatcher: %s", vm.Spec.Preference.Kind)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if preference != nil {
+		return &preference.Spec, nil
+	}
+
+	if clusterPreference != nil {
+		return &clusterPreference.Spec, nil
+	}
+
+	return nil, nil
```

Comment:
Do you think we should return an error here instead?
Something like `return nil, fmt.Errorf("No preference is found")`?

Reply:
Yup that's valid given L84 should return nil already if vm.Spec.Preference is nil, I'll address this now.

Reply:
Actually on reflection there shouldn't be a way for us to hit this return without already hitting an earlier error.